### PR TITLE
WC2-431: Decipher how to aggregate ETL based on current tableau

### PR DESCRIPTION
**Pull request by butofleury**
What problem is this PR solving? Explain here in one sentence.

Related JIRA tickets : [WC2-431](https://bluesquare.atlassian.net/browse/WC2-431)

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [ ] Is my code clear enough and well documented
- [ ] Are my typescript files well typed
- [ ] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [ ] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Changes

Aggregated data by period(month and year) and org unit.

## How to test

From the local instance with wfp coda database, run ETL script to generate data for:
- South Sudan: `docker compose run iaso manage etl_ssd `
- Nigeria: `docker compose run iaso manage etl_ng`

Then login on django admin and check on [Monthly statistics table](http://localhost:8081/admin/wfp/monthlystatistics/). You should see data aggregated by period and org unit. You can also filter the data per account(country).

## Print screen / video

[Administration-de-Wfp-Iaso.webm](https://github.com/user-attachments/assets/cd50e1a7-867b-481e-9464-c40e447cfcbe)

## Follow the Conventional Commits specification

The **merge message** of a pull request must follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) specification.

This convention helps to automatically generate release notes.

Use lowercase for consistency.

[Example](https://github.com/BLSQ/iaso/commit/8b8d7d3064138c1e57878f17b4eb922516ab0112):

```
fix: empty instance pop up

Refs: IA-3665
```

Note that the Jira reference is preceded by a _line break_.

Both the line break and the Jira reference are entered in the _Add an optional extended description…_ field.

[WC2-431]: https://bluesquare.atlassian.net/browse/WC2-431?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: plugins/wfp/aggregate_journeys.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+from itertools import groupby
+from operator import itemgetter
+
+
+class AGGREGATE_JOURNEY:
```
Comment (bramj) [inline plugins/wfp/aggregate_journeys.py]: Is there a reason why this is not CamelCase?

File: plugins/wfp/common.py [hunk 2]
```diff
@@ -637,3 +641,72 @@ def save_entity_journey(self, journey, beneficiary, record, entity_type):
...
+        monlthly_journey.admission_criteria = journey.get("admission_criteria")
+        monlthly_journey.given_sachet_rusf = journey.get("given_sachet_rusf")
+        monlthly_journey.given_sachet_rutf = journey.get("given_sachet_rutf")
+        monlthly_journey.given_quantity_csb = journey.get("given_quantity_csb")
+        monlthly_journey.exit_type = journey.get("exit_type")
+        monlthly_journey.account = account
+
+        monlthly_journey.save()
+
+    def journey_with_visit_and_steps_per_visit(self, account, program):
```
Comment (bramj) [inline plugins/wfp/common.py]: `programme`?

File: plugins/wfp/common.py [hunk 3]
```diff
@@ -637,3 +641,72 @@ def save_entity_journey(self, journey, beneficiary, record, entity_type):
...
+        aggregated_journey = []
```
Comment (bramj) [inline plugins/wfp/common.py]: Some variable names are pretty confusing. I think this would be better called `aggregated_journeys` to better reflect it's a list, what do you think?

File: plugins/wfp/common.py [hunk 4]
```diff
@@ -637,3 +641,72 @@ def save_entity_journey(self, journey, beneficiary, record, entity_type):
...
+                    Value("/"),
+                    Extract("visit__date", "month"),
+                    output_field=CharField(),
+                ),
+            )
+            .order_by("visit__id")
+        )
+        data_by_journey = groupby(list(journeys), key=itemgetter("visit__org_unit_id"))
+
+        for org_unit, journey in data_by_journey:
```
Comment (bramj) [inline plugins/wfp/common.py]: `journeys`

File: plugins/wfp/common.py [hunk 5]
```diff
@@ -637,3 +641,72 @@ def save_entity_journey(self, journey, beneficiary, record, entity_type):
...
+            visit_by_period = groupby(journey, key=itemgetter("period"))
```
Comment (bramj) [inline plugins/wfp/common.py]: `visits_by_period`

File: plugins/wfp/common.py [hunk 6]
```diff
@@ -637,3 +641,72 @@ def save_entity_journey(self, journey, beneficiary, record, entity_type):
...
+            assistance = {"rutf_quantity": 0, "rusf_quantity": 0, "csb_quantity": 0}
+            aggregated_journey = AGGREGATE_JOURNEY().group_by_period(
```
Comment (bramj) [inline plugins/wfp/common.py]: `aggregated_journeys`

File: plugins/wfp/models.py [hunk 7]
```diff
@@ -87,3 +87,20 @@ class Step(models.Model):
...
+    gender = models.CharField(max_length=8, choices=GENDERS, null=True, blank=True)
+    admission_criteria = models.CharField(max_length=255, choices=ADMISSION_CRITERIAS, null=True, blank=True)
+    admission_type = models.CharField(max_length=255, choices=ADMISSION_TYPES, null=True, blank=True)
+    exit_type = models.CharField(max_length=50, choices=EXIT_TYPES, null=True, blank=True)
+    nutrition_programme = models.CharField(max_length=255, choices=NUTRITION_PROGRAMMES, null=True, blank=True)
+    programme_type = models.CharField(max_length=255, choices=PROGRAMME_TYPE, null=True, blank=True)
+    number_visits = models.IntegerField(default=0)
+    given_sachet_rusf = models.FloatField(null=True, blank=True)
+    given_sachet_rutf = models.FloatField(null=True, blank=True)
+    given_quantity_csb = models.FloatField(null=True, blank=True)
```
Comment (bramj) [inline plugins/wfp/models.py:106]: Is there a reason these 3 quantity fields are floats? Intuitively I would think an integer makes more sense?

File: plugins/wfp/aggregate_journeys.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+                            "quantity_given", 0
+                        )
+                    if visit.get("assistance_type") in ["csb", "csb1", "csb2"]:
+                        assistance["csb_quantity"] = assistance.get("csb_quantity", 0) + visit.get("quantity_given", 0)
+                    if visit.get("visit__journey__exit_type") is not None:
+                        row["exit_type"] = visit.get("visit__journey__exit_type")
+                row["number_visits"] = len(all_visits)
+        return row
+
+    def group_by_period(self, visit_by_period, org_unit, all_journey, assistance):
```
Comment (bramj) [inline plugins/wfp/aggregate_journeys.py]: `visits_by_period`, `all_journeys`

File: plugins/wfp/common.py [hunk 9]
```diff
@@ -637,3 +641,72 @@ def save_entity_journey(self, journey, beneficiary, record, entity_type):
         journey.save()
 
         return journey
+
+    def save_monthly_journey(self, journey, account):
```
Comment (bramj) [inline plugins/wfp/common.py]: This feels like `journey` is an instance of `Journey`, but in reality it is a dict representing 1 monthly statistic. Maybe the naming can be improved?

File: plugins/wfp/aggregate_journeys.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+                    if nutrition_program is not None:
+                        row["nutrition_programme"] = nutrition_program
+                    row = self.group_visit_by_admission(current_visit, row, assistance)
+
+            row["org_unit"] = org_unit
+            row["given_sachet_rusf"] = assistance.get("rusf_quantity", 0)
+            row["given_sachet_rutf"] = assistance.get("rutf_quantity", 0)
+            row["given_quantity_csb"] = assistance.get("csb_quantity", 0)
+            all_journey.append(row)
+        return all_journey
```
Comment (bramj) [inline plugins/wfp/aggregate_journeys.py]: The code seems is working, but I find it pretty hard to follow to be honest. I think the main reason is that there are certain variables like `assistance` and `row` that are changed in place, and other methods that depend on these changes. I'm not sure if that's fixable, but I think we should avoid manipulating objects in place, and instead have functions return copies. We should strive for functions without side-effects, which makes the code easier to understand.

**Review by bramj — CHANGES_REQUESTED** [review]
Tested locally and working. However, I think variable naming could be better to improve the overall code readability. As well as (but more difficult) avoiding function side-effects.

File: plugins/wfp/models.py (hunk 7 above)
Comment (butofleury) [reply plugins/wfp/models.py:106]: > Is there a reason these 3 quantity fields are floats? Intuitively I would think an integer makes more sense?

 @bramj  there is some given assistance with float values in given quantity. 

image

**Review by bramj — APPROVED** [review]
LGTM :+1:
